src/main/java/org/apache/ibatis/builder/annotation/ProviderSqlSource.java

```diff
@@ -55,16 +58,34 @@ public ProviderSqlSource(Configuration configuration, Object provider) {
 
   /**
    * @since 3.4.5
+   * @deprecated Since 3.5.3, Please use the {@link #ProviderSqlSource(Configuration, Annotation, Class, Method)} instead of this.
+   * This constructor will remove at a future version.
    */
+  @Deprecated
   public ProviderSqlSource(Configuration configuration, Object provider, Class<?> mapperType, Method mapperMethod) {
+    this(configuration, provider instanceof Annotation ? (Annotation) provider :
+        (Annotation) Proxy.newProxyInstance(SelectProvider.class.getClassLoader(), new Class<?>[]{SelectProvider.class},
+          (proxy, method, args) -> {
+            if (method.getName().equals("annotationType")) {
+              return SelectProvider.class;
+            }
+            return provider.getClass().getMethod(method.getName()).invoke(provider);
+          }),
+      mapperType, mapperMethod);
+  }
```

Comment:
Can we just remove these old constructors?
I understand that you are trying to keep backward compatibility, but this code with proxy and the tests with `Object` look too unrealistic.
As this is an  internal class, I think it's OK to assume that the second parameter is `Annotation` and `mapperType` and `mapperMethod` are not `null`.

Reply:
If we treat the `ProviderSqlSource` as an internal class, we can delete unused constructors.

Reply:
Even if we keep the old constructors as deprecated, we can skip the `instanceof` check at least. A usage like `keepBackwardCompatibilityOnDeprecatedConstructor` is not intended.

Reply:
Fix via 4dbd30a3dd143bbdcb64d468af71642b4dce4315.